exporter/elasticsearchexporter/elasticsearch_bulk.go

```diff
@@ -207,18 +190,14 @@ func newBulkIndexer(logger *zap.Logger, client *elasticsearch7.Client, config *C
 			return nil, err
 		}
 		w := worker{
-			indexer:       bi,
-			items:         pool.items,
-			flushInterval: flushInterval,
-			flushTimeout:  config.Timeout,
-			flushBytes:    flushBytes,
-			logger:        logger,
-			stats:         &pool.stats,
+			indexer:      bi,
+			closeCh:      pool.closeCh,
+			flushTimeout: config.Timeout,
```

Comment:
A bit concerned about `timeout` configuration. This config option is currently used to configure flush timeout, however, I think we should eventually use the timeout_sender and the retry sender with the batch sender. The timeout_sender will also need a timeout setting and it will be better if we are prepared on how to resolve this. I think we have 2 options:

1. We can change the current meaning of the `timeout` configuration and introduce another option, maybe `flush_timeout`.
2. We can change how the timeout sender is configured by introducing a parent, something like `batchsender.timeout` maybe(?)

Reply:
Currently, flushTimeout is set to be config.Timeout. It is a client side timer to timeout a _single_ flush request. In the future, if we move to use timeout_sender, we will likely remove this client side timer, and to pass this config.Timeout to timeout_sender for its client side timer over `request.Export`. I don't see a need to introduce another option, as it has the same meaning to the user, which is "how long should a single bulk request take".

We have control over parsing of config, we could do something like this:
```go
func getRetryConfig(cf *Config) configretry.BackOffConfig {
	retryCfg := configretry.NewDefaultBackOffConfig()
	retryCfg.Enabled = cf.Retry.Enabled
	retryCfg.InitialInterval = cf.Retry.InitialInterval
	retryCfg.MaxInterval = cf.Retry.MaxInterval
	retryCfg.MaxElapsedTime = time.Duration(int64(cf.Retry.MaxRequests-1) * int64(cf.Retry.MaxInterval))
	return retryCfg
}
```

> We can change how the timeout sender is configured by introducing a parent, something like batchsender.timeout maybe(?)

I'm not sure if I understand. What are we trying to solve here?

Reply:
> In the future, if we move to use timeout_sender

We should already be using the timeout sender with the default config (of 5 seconds) -- the base exporter creates a default timeout sender ([ref](https://github.com/open-telemetry/opentelemetry-collector/blob/6115ef3498b8b234e01bb897096c53794d439228/exporter/exporterhelper/common.go#L265))

> In the future, if we move to use timeout_sender, we will likely remove this client side timer, and to pass this config.Timeout to timeout_sender for its client side timer over request.Export.

I agree with this however, I got confused since even now we would be using the TimeoutSender timeout. If this is the way forward then no worries from my side.

Reply:
> In the future, if we move to use timeout_sender, we will likely remove this client side timer, and to pass this config.Timeout to timeout_sender for its client side timer over request.Export.

We should be doing this in the current PR I think. This PR basically has a context passed to flush with timeout sender's default timeout (of 5 seconds) and then we derive another context based on flush timeout. This should mean that any flush timeout of > 5 seconds is ineffective (since the timeout sender will cancel the parent context)

Reply:
> the base exporter creates a default timeout sender ([ref](https://github.com/open-telemetry/opentelemetry-collector/blob/6115ef3498b8b234e01bb897096c53794d439228/exporter/exporterhelper/common.go#L265))

Nice catch. I missed that. In this PR I will disable timeout_sender to avoid it overriding the bulk indexer flush timeout.